src/Microsoft.OpenApi/Extensions/OpenApiSerializableExtensions.cs

```diff
@@ -86,6 +87,19 @@ public static void SerializeAsYaml<T>(this T element, Stream stream, OpenApiSpec
         /// <param name="specVersion">The Open API specification version.</param>
         public static void Serialize<T>(this T element, IOpenApiWriter writer, OpenApiSpecVersion specVersion)
             where T : IOpenApiSerializable
+        {
+            writer.Settings.SpecVersion = specVersion;
```

Comment:
I'm ok with making this change, but as you say it is breaking.  Currently we have a blend of two styles that overlap and is confusing.  We also have the weird issue that you could create a settings object that has format-JSON and pass it to the ctor of the OpenAPIYamlWriter.  And that's just weird.

---
In reply to: [194185318](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194185318) [](ancestors = 194185318)

Reply:
I'm starting to think having the Format property in the Settings is wrong.  The idea of have IOpenApiWriter was so that the output formats was extensible.  By creating the Format enum we have broken that extensibility.

---
In reply to: [194188553](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194188553) [](ancestors = 194188553,194185318)

Reply:
To make it sound less severe, I would argue this current change is also breaking. It's modifying the writer object.

Do we really need the Settings to be in the Writer? Would things be more consistent if we remove it and disallow Serialize<T>(this T element, IOpenApiWriter writer)?



---
In reply to: [194188553](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194188553) [](ancestors = 194188553,194185318)

Reply:
I need it specifically to support the X-MS-Paths extension.  I have created a custom extension that reuses the PathItem object within the extension.  The IOpenApiExtension interface has a Write method that only receives the writer object, so I need to be able to call pathItem.Serialize(writer) and it maintain the format and spec version of the document is is contained in.

---
In reply to: [194189250](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194189250) [](ancestors = 194189250,194188553,194185318)

Reply:
- I think you are right about the format. If we force IOpenApiWriter to have a format, then it loses that extensibility. 

- I'm unsure whether we need the specVersion in the writer. If we leave it out, would things be simpler?

Reply:
I think the explicit parameters should just be helpers that hide the creation of the appropriate Writer and Settings object.

---
In reply to: [194189876](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194189876) [](ancestors = 194189876,194189250,194188553,194185318)

Reply:
I see. What if we change the interface for IOpenApiExtension to Write(IOpenApiWriter writer, OpenApiSpecVersion specVersion) instead? I'm exploring this route because it seems to me that the OpenApiWriter objects should only care about the format (JSON, YAML, etc.) itself and not what it's writing (v2, v3, ...)

---
In reply to: [194189876](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194189876) [](ancestors = 194189876,194189250,194188553,194185318)

Reply:
The other option you mentioned above probably works too... I guess the hard problem there will be how to make it clear the "writer" object given will be modified.

---
In reply to: [194190967](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194190967) [](ancestors = 194190967,194189876,194189250,194188553,194185318)

Reply:
Well, we can have two distinct approaches: 1) Give us a writer and we won't touch it.  2) Give us explicit properties and we'll create the writer for you.

---
In reply to: [194191141](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194191141) [](ancestors = 194191141,194190967,194189876,194189250,194188553,194185318)

Reply:
It is not impossible that in future versions we may allow certain YAML constructs that were not allowed in earlier versions.  In theory, V3 is Yaml 1.2 whereas V2 was technically Yaml 1.1.  That different was not really specified very explicitly but technically it is the case.

---
In reply to: [194192111](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194192111) [](ancestors = 194192111,194191141,194190967,194189876,194189250,194188553,194185318)

Reply:
I'm going to take a crack at cleaning this up.  Let' me know what you think.  Whatever we do is going to end up being some level of breaking at this point.

---
In reply to: [194192408](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194192408) [](ancestors = 194192408,194192111,194191141,194190967,194189876,194189250,194188553,194185318)

Reply:
- Agree we can't avoid breaking something now.

- The YAML version issue: This seems like something we need to fix with having a YamlV1_2Writer or something like that. Otherwise, we are tying the specVersion and format together.

- I'm okay with your approach above "1) Give us a writer and we won't touch it.  2) Give us explicit properties and we'll create the writer for you." This seems to be the cleanest among the options we mentioned.

- We will still need to remove the Format from the IOpenApiWriter like you mentioned - not sure what other problems that can lead to.

---
In reply to: [194192111](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194192111) [](ancestors = 194192111,194191141,194190967,194189876,194189250,194188553,194185318)

Reply:
I just removed the Serialize method that takes a writer and a spec version, and I removed the Format from the Settings object, and nothing broke and no tests failed.  I guess they weren't that useful after all!

---
In reply to: [194193451](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194193451) [](ancestors = 194193451,194192111,194191141,194190967,194189876,194189250,194188553,194185318)

Reply:
I need to think some more about the combined spec/format Writer concern that you have.  I'll also look at what is the impact of passing the version as a parameter to IOpenApiExtension.  If we did that we would not need a the settings object at all any more.  

---
In reply to: [194193664](https://github.com/Microsoft/OpenAPI.NET/pull/269#discussion_r194193664) [](ancestors = 194193664,194193451,194192111,194191141,194190967,194189876,194189250,194188553,194185318)